src/libqof/qof/gnc-numeric.cpp

```diff
@@ -21,46 +21,501 @@
 #include <string.h>
-#ifdef __cplusplus
+
+#include "qof.h"
 }
-#endif
+
 #include <stdint.h>
+#include <boost/regex.hpp>
+#include <sstream>
+#include <cstdlib>
 
-#include "gnc-numeric.h"
+#include "gnc-numeric.hpp"
 #include "gnc-rational.hpp"
 
-using GncNumeric = GncRational;
-
-static const gint64 pten[] = { 1, 10, 100, 1000, 10000, 100000, 1000000,
-			       10000000, 100000000, 1000000000,
-			       INT64_C(10000000000), INT64_C(100000000000),
-			       INT64_C(1000000000000), INT64_C(10000000000000),
-			       INT64_C(100000000000000),
-			       INT64_C(10000000000000000),
-			       INT64_C(100000000000000000),
-			       INT64_C(1000000000000000000)};
+static QofLogModule log_module = "qof";
+static const int64_t pten[] = { 1, 10, 100, 1000, 10000, 100000, 1000000,
+                               10000000, 100000000, 1000000000,
+                               INT64_C(10000000000), INT64_C(100000000000),
+                               INT64_C(1000000000000), INT64_C(10000000000000),
+                               INT64_C(100000000000000),
+                               INT64_C(10000000000000000),
+                               INT64_C(100000000000000000),
+                               INT64_C(1000000000000000000)};
 #define POWTEN_OVERFLOW -5
 
-static inline gint64
-powten (int exp)
+int64_t
+powten (unsigned int exp)
+{
+    if (exp > 17)
+        exp = 17;
+    return pten[exp];
+}
+
+GncNumeric::GncNumeric(GncRational rr)
+{
+    /* Can't use isValid here because we want to throw different exceptions. */
+    if (rr.num().isNan() || rr.denom().isNan())
+        throw std::underflow_error("Operation resulted in NaN.");
+    if (rr.num().isOverflow() || rr.denom().isOverflow())
+        throw std::overflow_error("Operation overflowed a 128-bit int.");
+    if (rr.num().isBig() || rr.denom().isBig())
+    {
+        GncRational reduced(rr.reduce());
+        rr = reduced.round_to_numeric(); // A no-op if it's already small.
+    }
+    m_num = static_cast<int64_t>(rr.num());
+    m_den = static_cast<int64_t>(rr.denom());
+}
+
+GncNumeric::GncNumeric(double d) : m_num(0), m_den(1)
+{
+    if (isnan(d) || fabs(d) > 1e18)
+    {
+        std::ostringstream msg;
+        msg << "Unable to construct a GncNumeric from " << d << ".\n";
+        throw std::invalid_argument(msg.str());
+    }
+    constexpr auto max_denom = INT64_MAX / 10;
+    auto logval = log10(fabs(d));
+    int64_t den;
+    if (logval > 0.0)
+        den = powten(18 - static_cast<int>(floor(logval) + 1.0));
+    else
+        den = powten(17);
+    auto num = static_cast<int64_t>(floor(static_cast<double>(den) * d));
+
+    if (num == 0)
+        return;
+    GncNumeric q(num, den);
+    auto r = q.reduce();
+    m_num = r.num();
+    m_den = r.denom();
+}
+
+using boost::regex;
+using boost::smatch;
+using boost::regex_search;
+GncNumeric::GncNumeric(const std::string& str, bool autoround)
+{
+    static const std::string numer_frag("(-?[0-9]+)");
```

Comment:
I tried to get operator>> to accept thousands separators, but it looks like the llvm libstdc++ doesn't support that, at least as of what Apple included in their latest Xcode.

For now I think we continue to use xaccPrintNumeric and xaccParseNumeric and look at merging that functionality into GncNumeric later when time permits.